### PR TITLE
Add basic game and possession class

### PR DESCRIPTION
**Pull request by jalyna**
@PragTob Please have a look. This is just the basic setup of the Game and Possession Class. The logic of what kind of possessions have to be generated is still missing and how they are executed. Also the logic how long one possession took. In the future it makes sense that every possession generates its own game log line.
- Created squad factory
- Created a Game class that generates possessions (with `run`) as long as the current game length is lower than 12*4 minutes
- A game has to be initialized with 2 squads
- A possession has to be initialized with `owner_squad` and `opponent_squad`

### CODE REVIEW

File: lib/bb_matchengine.rb [hunk 1]
```diff
@@ -1,8 +1,12 @@
-require "bb_matchengine/version"
+require 'bb_matchengine/version'
```
Comment (PragTob) [inline lib/bb_matchengine.rb:1]: :+1:

File: lib/bb_matchengine/game.rb [hunk 2]
```diff
@@ -0,0 +1,36 @@
+module BBMatchengine
+  class Game
+
+    # http://en.wikipedia.org/wiki/Basketball#Playing_regulations (NBA)
+    QUARTER_LENGTH = 12 * 60
```
Comment (PragTob) [inline lib/bb_matchengine/game.rb]: We wanna do European rules (10 \* 60) - manager games are more popular in Europe plus the European league system is closer to what will be built than the NBA one.

File: lib/bb_matchengine/game.rb [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+      last_possession.public_send(squad)
+    end
+
+    private
+    def last_possession
+      possessions.last
+    end
+
+    def current_length
+      possessions.inject(0) { |sum, p| sum + p.length }
```
Comment (PragTob) [inline lib/bb_matchengine/game.rb]: I'd keep a variable keeping track of the current_time/length in the game class to avoid frequent iterations

**Comment by PragTob** [comment]
Looks good overall!

From a design perspect, though I'd opt for a more procedural way of determining possesions. E.g. more in a loop. It fits better with my mental model... plus what kind of possesion it is is only determined at the end. I think plenty of objects will emerge here nontheless.

I.e. (hacked down pseudo code):

``` ruby
until game_end?
  play_possession
  switch_teams
end

def play_possession
  until possesion_end?
    perform_action
    wait_random_time # simulates some time loss
  end
end

def perform_action
  current_player.decide_on_action # something/magic
  # some if statement, or wait the above might return an Action object or something.. that might be an action (like Pass/shoot/keep the ball)
  # anyways, here are some actions that can result
  player.shoot # basket/no basket (latter one results in rebound)
  player.pass # turn over or new player gets the ball
end
```

Seeing that, a Possesion class maybe IS a good idea (always great to write stuff out). But I don't think we need the subclasses as that is determined during the possession. 

However, there only ever is one possesion at a time and I don't think it makes too much sense to keep track of the possesions (could be dead wrong here). I'd rather publish events that relevant services could subscribe to and perform actions accordingly. E.g. a `MadeShotEvent` is interesting to both the statistics service (+ points for team and player) and the game log service (generating the log). While a PassEvent is only relevant for the GameLog service (unless we want to keep stats over all passes, which is unusual/never seen it).

**Comment by jalyna** [comment]
I'm not sure if it doesn't make sense to keep posSessions. I would prefer having a game with memories, for more strategic options in the future. As I said events would be also necessary in the future.

But I think I got the possession thing wrong. Maybe it makes more sense if a game has possessions which each have different actions, maybe with player assigned?

e.g.:

``` ruby
class Game
  def run
    until game_end?
      @possessions << Possession.new(current_owner, current_opponent)
      @current_length += last_possession.length
      switch_teams
    end
  end
end

class Possession
  attr_reader :actions

  def run
    until possesion_end?
      # ... as you mentioned but I think it should be run
    end
  end
end

# Maybe Action should be something polymorphic?
class Action
  attr_reader :player

  def run
    # all magic logic goes here
  end
end
```

**Comment by PragTob** [comment]
I don't think that most strategic options require a memory. Most strategic options are dependent on what is currently going on, such as the score, players on court, team fouls, current opponent, opponent strategy, the current time etc...

It might become useful in the future I'd want to build a simple prototype for now and this can be added if in the future this feature is really needed. So far we don't even have a strategy or tactics settings.

As for possesions, I think players can conduct the actions - otherwise they are just value objects but the basic logic for calculating a `pass` value or something could go to them.

**Pull request merged**